### PR TITLE
HPCC-16323 Insure uslaves.start file present for killing slaves

### PR DESCRIPTION
**Pull request by Michael-Gardner**
Signed-off-by: Michael Gardner michael.gardner@lexisnexis.com

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
@mckellyln building now and testing on my system to make sure it's all correct.  If you could look over it quickly and perhaps test out the change on your cluster I'd appreciate it.

File: initfiles/bin/init_thor [hunk 1]
```diff
@@ -100,14 +100,13 @@ kill_slaves()
     if [[ "$localthor" = "true" ]]; then
         $deploydir/init_thorslave stop localhost $THORMASTER $THORMASTERPORT $LOG_DIR $instancedir $deploydir $THORNAME $PATH_PRE $logredirect
     else
-        # we want to kill only slaves that have already been started in run_thor
-        if [[ -r $instancedir/uslaves.start ]]; then
-            nslaves=$(cat $instancedir/uslaves.start 2> /dev/null | wc -l)
-            $deploydir/frunssh $instancedir/uslaves.start "/bin/sh -c '$deploydir/init_thorslave stop %a $THORMASTER $THORMASTERPORT $LOG_DIR $instancedir $deploydir $THORNAME $PATH_PRE $logredirect'" -i:$SSHidentityfile -u:$SSHusername -pe:$SSHpassword -t:$SSHtimeout -a:$SSHretries -n:$nslaves 2>&1
-        fi
+        daliadmin server=$DALISERVER dfsgroup ${groupName} slaves
```
Comment (mckellyln) [inline initfiles/bin/init_thor]: I think I would use the uslaves.start file if it exists, otherwise do the daliadmin cmd to recreate it.

**Review by mckellyln — COMMENTED** [review]
@Michael-Gardner just one comment.

File: initfiles/bin/init_thor [hunk 2]
```diff
@@ -100,14 +100,15 @@ kill_slaves()
     if [[ "$localthor" = "true" ]]; then
         $deploydir/init_thorslave stop localhost $THORMASTER $THORMASTERPORT $LOG_DIR $instancedir $deploydir $THORNAME $PATH_PRE $logredirect
     else
-        # we want to kill only slaves that have already been started in run_thor
-        if [[ -r $instancedir/uslaves.start ]]; then
-            nslaves=$(cat $instancedir/uslaves.start 2> /dev/null | wc -l)
-            $deploydir/frunssh $instancedir/uslaves.start "/bin/sh -c '$deploydir/init_thorslave stop %a $THORMASTER $THORMASTERPORT $LOG_DIR $instancedir $deploydir $THORNAME $PATH_PRE $logredirect'" -i:$SSHidentityfile -u:$SSHusername -pe:$SSHpassword -t:$SSHtimeout -a:$SSHretries -n:$nslaves 2>&1
+        if [[ ! -e "${instancedir}/uslaves.start" ]]; then
```
Comment (Michael-Gardner) [inline initfiles/bin/init_thor]: Added conditional check for uslaves.start before we do daliadmin, so we don't overwrite if it exists.

**Comment by richardkchapman** [comment]
@jakesmith I'm getting a feeling of deja-vu on this... wasn't there a similar fix applied recently?

File: initfiles/bin/init_thor [hunk 3]
```diff
@@ -198,7 +199,7 @@ while [[ 1 ]]; do
         case $errcode in
         # TEC_Clean
         0)  log "Thormaster ($thorpid) Exited cleanly"
-            rm -f $instancedir/uslaves.start $PID_NAME $INIT_PID_NAME > /dev/null 2>&1
+            rm -f $PID_NAME $INIT_PID_NAME > /dev/null 2>&1
```
Comment (mckellyln) [inline initfiles/bin/init_thor]: I think we should leave in the remove file here since it was stopped clean and not restarted.

**Review by mckellyln — COMMENTED** [review]
One more comment about leaving in the rm when stopped cleanly.

File: initfiles/bin/init_thor [hunk 4]
```diff
@@ -198,7 +199,7 @@ while [[ 1 ]]; do
         case $errcode in
         # TEC_Clean
         0)  log "Thormaster ($thorpid) Exited cleanly"
-            rm -f $instancedir/uslaves.start $PID_NAME $INIT_PID_NAME > /dev/null 2>&1
+            rm -f $PID_NAME $INIT_PID_NAME $instancedir/uslaves.start > /dev/null 2>&1
```
Comment (Michael-Gardner) [inline initfiles/bin/init_thor]: Added the removal back in like suggested

**Review by mckellyln — APPROVED** [review]

**Comment by richardkchapman** [comment]
@jakesmith Can you comment?

**Comment by jakesmith** [comment]
>@jakesmith I'm getting a feeling of deja-vu on this... wasn't there a similar fix applied recently?

@richardkchapman - you may be thinking of : https://track.hpccsystems.com/browse/HPCC-16120

**Comment by jakesmith** [comment]
@Michael-Gardner - I am still not sure why this change necessary.
IOW - how does uslaves.start go missing?

The idea (pre this commit) was that uslaves.start was created (and left) after a successful start and used at exit time.
It was not recreated (via daliadmin), in case someone had pushed a reconfigured the environment whilst thor was still up.

I notice in master, as it stands, that if there are any ssh errors during the kill_slaves frunssh call it will abort and log the frunssh errors.
Whereas in 5.6.8 - it would have continued, deleting uslaves.start.
So for if (e.g.) a machine was uncontactable during kill, it could leave slaves up and running and still be there as it tries to start back up.

**Comment by jakesmith** [comment]
@Michael-Gardner - I am not sure of:

a) the original reason the uslaves.start is missing (is above the reason suspected) ?
b) The significance of master's handling of frunssh errors in kill_slaves difference.


If it's a temporary inability to contact a slave that leads to slave process(es) being left up and causing conflict.
I think it would be better to:
1) Report frunssh errors from kill and abort.
2) Leave uslaves.start (uslaves in master) in place if any errors.
3) If uslaves.start present at next start, use it to try and kill off any stray slaves.

**Comment by Michael-Gardner** [comment]
Closing after talking with Mark.  Will do new PR for 6.2 regarding logging check for uslaves.

**Pull request closed without merging**